### PR TITLE
fix(codegen): loop combing should not reuse same count value for different columns

### LINKED ISSUE #1927: Get incorrect result for min/max/count with null data which may be related to loop-combining codegen
Now we may get incorrect result for methods of min/max/count with null data. Here is simple case which can be run in offline directly.

```
package com._4paradigm.openmldb.batchjob.tools.tobe.loaddata

import com._4paradigm.openmldb.batch.api.OpenmldbSession
import org.apache.spark.sql.{Row, SparkSession}
import org.apache.spark.sql.types.{IntegerType, StringType, StructField, StructType, TimestampType}
import java.sql.Timestamp

object SameWindowIssueLocal {

  def main(args: Array[String]): Unit = {

    val spark = SparkSession.builder().master("local")
      .getOrCreate()

    val sess = new OpenmldbSession(spark)

    val data = Seq(
      Row(1, "g1", Timestamp.valueOf("2021-02-01 13:59:58"), 1, -3),
      Row(2, "g1", Timestamp.valueOf("2021-02-01 13:59:59"), null, null),
      Row(3, "g1", Timestamp.valueOf("2021-02-01 14:00:00"), 2, -1),
      Row(4, "g2", Timestamp.valueOf("2021-02-01 14:00:00"), null, -1),
      Row(5, "g2", Timestamp.valueOf("2021-02-01 14:00:01"), null, -2),
      Row(6, "g3", Timestamp.valueOf("2021-02-01 13:54:02"), 1, 1),
    )
    val schema = StructType(List(
      StructField("row_id", IntegerType),
      StructField("key1", StringType),
      StructField("ts", TimestampType),
      StructField("l1", IntegerType),
      StructField("l2", IntegerType)
    ))
    val df = spark.createDataFrame(spark.sparkContext.makeRDD(data), schema)
    sess.registerTable("t1", df)

    val sqlText =
      """
        |select
        |    `row_id` as row_id_1,
        |    max(`l1`) over t6_2_key1_ts_0s_5s_10 as t6_2_l1_window_max_8,
        |    max(`l2`) over t6_2_key1_ts_0s_5s_10 as t6_2_l2_window_max_9
        |from
        |    `t1`
        |    window t6_2_key1_ts_0s_5s_10 as (partition by `key1` order by `ts` rows_range between 5s open preceding and 0s preceding MAXSIZE 10);
        |""".stripMargin

    val outputDf = sess.sql(sqlText)
    outputDf.show()

    val sqlText2 =
      """
        |select
        |    `row_id` as row_id_1,
        |    max(`l2`) over t6_2_key1_ts_0s_5s_10 as t6_2_l2_window_max_9
        |from
        |    `t1`
        |    window t6_2_key1_ts_0s_5s_10 as (partition by `key1` order by `ts` rows_range between 5s open preceding and 0s preceding MAXSIZE 10);
        |""".stripMargin

    val outputDf2 = sess.sql(sqlText2)
    outputDf2.show()
  }

}
``` 

If we only select the non-null column, the result is correct.

```
+--------+--------------------+
|row_id_1|t6_2_l2_window_max_9|
+--------+--------------------+
|       4|                  -1|
|       5|                  -1|
|       6|                   1|
|       1|                  -3|
|       2|                  -3|
|       3|                  -1|
+--------+--------------------+
```

If we select two columns and one of these columns has null data, the result is incorrect and make all result is null.

```
+--------+--------------------+--------------------+
|row_id_1|t6_2_l1_window_max_8|t6_2_l2_window_max_9|
+--------+--------------------+--------------------+
|       4|                null|                null|
|       5|                null|                null|
|       6|                   1|                   1|
|       1|                   1|                  -3|
|       2|                   1|                  -3|
|       3|                   2|                  -1|
+--------+--------------------+--------------------+
```

If we change the method from `max` to `add` which may not use loop-combining CodeGen, the result will be correct. So it may be the issue of handling null data in loop-combining CodeGen.

### LINKED ISSUE #1951: `window rows between 0 preceding and 2 preceding` should report compile error
**Bug Description**

the behavior of sql parser should be the same with spark-sql.
in spark, query like this should result compile time error
```sql
spark-sql> select x, sum(y) over (partition by z order by x rows between 0 preceding and 2 preceding) from values ((1, 1, 2)), ((2, 2, 2)) as tab(x, y, z);
Error in query: cannot resolve 'ROWS BETWEEN 0 PRECEDING AND 2 PRECEDING' due to data type mismatch: The lower bound of a window frame must be less than or equal to the upper bound; line 1 pos 22;
'Project [x#12, unresolvedalias(sum(y#13) windowspecdefinition(z#14, x#12 ASC NULLS FIRST, specifiedwindowframe(RowFrame, -0, -2)), None)]
+- SubqueryAlias tab
   +- LocalRelation [x#12, y#13, z#14]

spark-sql> select x, sum(y) over (partition by z order by x rows between 2 following and 0 following) from values ((1, 1, 2)), ((2, 2, 2)) as tab(x, y, z);
Error in query: cannot resolve 'ROWS BETWEEN 2 FOLLOWING AND 0 FOLLOWING' due to data type mismatch: The lower bound of a window frame must be less than or equal to the upper bound; line 1 pos 22;
'Project [x#20, unresolvedalias(sum(y#21) windowspecdefinition(z#22, x#20 ASC NULLS FIRST, specifiedwindowframe(RowFrame, 2, 0)), None)]
+- SubqueryAlias tab
   +- LocalRelation [x#20, y#21, z#22]
```
spark always expect `start_frame <= end_frame`

currently we just treat those window as window without data

**Expected Behavior**

**Steps to Reproduce**

1. 
2. 
3. 
4.

### PR DESCRIPTION
**Pull request by aceforeverd**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

- fix: #1927 
- fix: #1951 
- partial #1586 

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

### CODE REVIEW

**Comment by aceforeverd** [comment]
[bugfix: sql engine related bugs](https://trello.com/c/Gu5DKwue/104-bugfix-sql-engine-related-bugs)

**Review by tobegit3hub — APPROVED** [review]
LGTM

File: hybridse/src/codegen/aggregate_ir_builder.cc [hunk 1]
```diff
@@ -236,34 +232,25 @@ class StatisticalAggGenerator {
...
     void GenInitState(::llvm::IRBuilder<>* builder) {
         for (size_t i = 0; i < col_num_; ++i) {
+            // any of min/max/avg/sum/count need the count info
+            count_state_[i] = GenCountInitState(builder);
+
             if (!sum_idxs_[i].empty()) {
                 sum_states_[i] = GenSumInitState(builder);
             }
             if (!avg_idxs_[i].empty()) {
                 if (col_type_ == ::hybridse::node::kDouble) {
```
Comment (dl239) [inline hybridse/src/codegen/aggregate_ir_builder.cc:242]: why only judge `kDouble` here.

Comment (aceforeverd) [reply hybridse/src/codegen/aggregate_ir_builder.cc:242]: the idea is generally:
- if column type is double, `avg` and `sum` can share the sum value

on output, avg requires `sum` and `count` to be float

**Review by zhanghaohit — APPROVED** [review]
LGTM

**Review by dl239 — APPROVED** [review]

**Pull request merged**